### PR TITLE
[4.1] [Fix] display translatable field in relationship

### PR DESCRIPTION
**Pull request by t-prod**
Able to display when you have a translatable field the value of app()->getLocale()
Check https://github.com/Laravel-Backpack/CRUD/issues/2647

### CODE REVIEW

**Comment by pxpm** [comment]
Hey @t-prod 

Thanks for putting up a solution for this. 

I agree with you that we need this change to parse translatable attributes, but I think we need to parse translatable attributes ALL the time this function return the attribute, and you just accounted for one return scenario.

My proposal to solve this would be something along this lines:

```php
                   public function getRelatedEntriesAttributes($model, $relationString, $attribute)
    {
        $endModels = $this->getRelatedEntries($model, $relationString);
        $attributes = [];
        foreach ($endModels as $model => $entries) {
            $model_instance = new $model();
            $modelKey = $model_instance->getKeyName();
            if (is_array($entries) && ! isset($entries[$attribute])) {
                foreach ($entries as $entry) {
                    $attributes[$entry[$modelKey]] = $this->parseTranslatableAttributes($model_instance, $attribute, $entry[$attribute]);
                }
            } elseif (is_array($entries) && isset($entries[$attribute])) {
                $attributes[$entries[$modelKey]] = $this->parseTranslatableAttributes($model_instance, $attribute, $entry[$attribute]);
            } elseif ($entries->{$attribute}) {
                $attributes[$entries->{$modelKey}] =$this->parseTranslatableAttributes($model_instance, $attribute, $entries->{$attribute});
            }
        }

        return $attributes;
    }
```

And create a new function that will do all the checks for translatable attributes and return the attribute accordingly.

```php
 public function parseTranslatableAttributes($model, $attribute, $value) {
        $translatableAttributes = $model->getTranslatableAttributes();
        if(empty($translatableAttributes)) {
            return $value;
        }
        if(!in_array($attribute, $translatableAttributes)) {
            return $value;
        }
        $decodedAttribute = json_decode($value, true);
        if(is_array($decodedAttribute) && !empty($decodedAttribute)) {
            if(isset($decodedAttribute[app()->getLocale()])) {
                return $decodedAttribute[app()->getLocale()];
            }else{
                return Arr::first($decodedAttribute);
            }          
        }

        return $value;
    }
```

I haven't tested this yet, but I think this will get us better cover, what do you think? 

Do you mind testing this code in your side and let me know how it goes? 

Best,
Pedro

**Comment by t-prod** [comment]
Hey @pxpm 

Thanks for your answer you right I'm only test this in a single scenario. I'll try with your advices and tell you tonight ;)

Thanks again

**Comment by t-prod** [comment]
It's working with a little fix by detecting $value is an array before json_encode.

**Comment by t-prod** [comment]
Hey @pxpm 

If you can accept my PR it will be great. I foudn an other issue I'll post it in issues.

File: src/app/Library/CrudPanel/CrudPanel.php [hunk 1]
```diff
@@ -348,21 +348,58 @@ public function getRelatedEntriesAttributes($model, $relationString, $attribute)
...
                 foreach ($entries as $entry) {
-                    $attributes[$entry[$modelKey]] = $entry[$attribute];
+                    $attributes[$entry[$modelKey]] = $this->parseTranslatableAttributes($model_instance, $attribute, $entry[$attribute]);
                 }
             } elseif (is_array($entries) && isset($entries[$attribute])) {
-                $attributes[$entries[$modelKey]] = $entries[$attribute];
+                $attributes[$entries[$modelKey]] = $this->parseTranslatableAttributes($model_instance, $attribute, $entries[$attribute]);
             } elseif ($entries->{$attribute}) {
-                $attributes[$entries->{$modelKey}] = $entries->{$attribute};
+                $attributes[$entries->{$modelKey}] =$this->parseTranslatableAttributes($model_instance, $attribute, $entries->{$attribute});
```
Comment (pxpm) [inline src/app/Library/CrudPanel/CrudPanel.php:360]: I would just give this a space `= $this ...`

**Comment by pxpm** [comment]
Dunno why styleCI is failing, could you please correct the space and send the commit again? 

StyleCI wants to space some stuff and add a period at the end of the sentence, but it's not doing that automatically, maybe @tabacitu can have a look. 

Apart from that this code fixes when you have a belongTo category, and category name is a translatable attribute.

**Comment by t-prod** [comment]
Hi @pxpm,

As you can see that's OK. Thanks to merge this PR and thanks for your help ;)

**Comment by t-prod** [comment]
Hi,

Is eveything all right ? Could you do it please ?

Thanks

**Comment by tabacitu** [comment]
Is this ready to merge @pxpm ? Does it have your stamp of approval?

**Comment by t-prod** [comment]
Hi,

If you can merge this please it's quite urgent ;) Thanks

**Comment by pxpm** [comment]
@t-prod sorry for taking time merge this. 

We are polishing stuff for 4.1. 

You can setup this in your project, when we merge it, just update the project and your changes will be lost and updated with backpack code. 

Thanks for your understanding.

**Comment by pxpm** [comment]
@tabacitu from my tests this is indeed needed for related translatable attributes. 

Bare with me:

`ProductModel` has `name` translatable.
`MonsterModel` has a `BelongsTo` relation with `ProductModel` 

When displaying the `ProductModel name` in `MonsterCrudList` without this, we will endup with an array in the `select` column instead of a string. 

It will fail because of this bit: `$text = Str::limit($text, $column['limit'], '[...]');` in select.blade.php

Str::limit expects a string not an array like:

```php
['en' => 'Product name']
```

Screenshot_2

Also while searching, select.blade.php is the only place where we call this function.

Also our multiple checks makes sure we only handle attributes that the model has translatable trait and the corresponding attribute is translatable. 

```php
        if (! method_exists($model, 'isTranslatableAttribute')) {
            return $value;
        }

        if (! $model->isTranslatableAttribute($attribute)) {
            return $value;
        }
```
## Second possible scenario:

We can also do the change in select.blade.php eg:

```php
foreach ($attributes as $key => $text) {
        if(is_array($text)) {
            if(!empty($text[app()->getLocale()])) {
                $text = $text[app()->getLocale()];
            }else{
                $text = Arr::first($text);
            }
        }
        $text = Str::limit($text, $column['limit'], '[...]');
    }
```
**But this needs change in both: Line 8 and Line 26.** 

We can extract that bit of code into some function like:

```php
public function parseTranslatableAttribute($model, $attr, $value) {
if(!method_exists($model, 'isTranslatableAttribute')) {
//we need a string, not an array.
            return json_encode($value);
        }

        if(!(new $model)->isTranslatableAttribute($attr)) {
//we need a string, not an array.
            return json_encode($value);
        }
    if(!empty($value[app()->getLocale()])) {
        return $value[app()->getLocale()];
    }else{
        return Arr::first($value);
    }
}
```
And then call it in select.blade.php

```php
foreach ($attributes as $key => $text) {
        if(is_array($text)) {
           $text = $crud->parseTranslatableAttribute($column['model'], $column['attribute'], $text);
        }
        $text = Str::limit($text, $column['limit'], '[...]');
    }
```

I think my last proposal is cleaner and in case you need to override some logic it's easier to extend the column than the crud panel method. 

@tabacitu Can you let me know your thoughts on this one?

**Comment by tabacitu** [comment]
thanks for taking a look at this @pxpm . I agree with you - it probably makes sense to move this to the column, since it's only used there. But then it would bloat up the column...

I decided to merge this as-is, since it works perfectly. Indeed - that might not be the right place for this code, but it's a small problem that can be easily refactored in 4.1.x, and it's the way it's always been, so it's not a new problem 😄 

Thanks again for submitting the PR @t-prod - very clean. Keep 'em coming.

Cheers!

**Pull request merged**